plugins/hls-notes-plugin/src/Ide/Plugin/Notes.hs

```diff
@@ -0,0 +1,142 @@
+                                                   SMethod (SMethod_TextDocumentDefinition))
+import           Language.LSP.Protocol.Types
+import qualified Language.LSP.Server              as LSP
+import           Language.LSP.VFS                 (VirtualFile (..))
+import           Text.Regex.TDFA                  (Regex, caseSensitive,
+                                                   defaultCompOpt,
+                                                   defaultExecOpt,
+                                                   makeRegexOpts, matchAllText)
+
+data Log
+    = LogShake Shake.Log
+    | LogNotesFound NormalizedFilePath [(Text, Position)]
+    deriving Show
+
+data GetNotesInFile = MkGetNotesInFile
+    deriving (Show, Generic, Eq, Ord)
+    deriving anyclass (Hashable, NFData)
+type instance RuleResult GetNotesInFile = HM.HashMap Text Position
+
+data GetNotes = MkGetNotes
+    deriving (Show, Generic, Eq, Ord)
+    deriving anyclass (Hashable, NFData)
+type instance RuleResult GetNotes = HashMap Text (NormalizedFilePath, Position)
+
+instance Pretty Log where
+    pretty = \case
+        LogShake l -> pretty l
+        LogNotesFound file notes ->
+            "Found notes in " <> pretty (show file) <> ": ["
+            <> pretty (intercalate ", " (fmap (\(s, p) -> "\"" <> s <> "\" at " <> T.pack (show p)) notes)) <> "]"
+
+{-
+The first time the user requests a jump-to-definition on a note reference, the
+project is indexed and searched for all note definitions. Their location and
+title is then saved in the HLS database to be retrieved for all future requests.
+-}
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId = (defaultPluginDescriptor plId "Provides goto definition support for GHC-style notes")
+    { Ide.Types.pluginRules = findNotesRules recorder
+    , Ide.Types.pluginHandlers = mkPluginHandler SMethod_TextDocumentDefinition jumpToNote
+    }
+
+findNotesRules :: Recorder (WithPriority Log) -> Rules ()
+findNotesRules recorder = do
+    defineNoDiagnostics (cmapWithPrio LogShake recorder) $ \MkGetNotesInFile nfp -> do
+        findNotesInFile nfp recorder
+
+    defineNoDiagnostics (cmapWithPrio LogShake recorder) $ \MkGetNotes _ -> do
+        targets <- toKnownFiles <$> useNoFile_ GetKnownTargets
+        definedNotes <- catMaybes <$> mapM (\nfp -> fmap (HM.map (nfp,)) <$> use MkGetNotesInFile nfp) (HS.toList targets)
+        pure $ Just $ HM.unions definedNotes
+
+jumpToNote :: PluginMethodHandler IdeState Method_TextDocumentDefinition
+jumpToNote state _ param
+    | Just nfp <- uriToNormalizedFilePath uriOrig
+    = do
+        let Position l c = param ^. L.position
+        contents <- fmap _file_text . err "Error getting file contents"
+            =<< lift (LSP.getVirtualFile uriOrig)
+        line <- err "Line not found in file" (Rope.lines contents ^? ix (fromIntegral l))
```

Comment:
Rather than getting all rope lines it might be faster to use 
Rope.splitAt twice as it's done e.g. here:
https://github.com/haskell/haskell-language-server/blob/50923e5c790e9c55c2b9b1bfcc8e78f7169f8ea3/ghcide/src/Development/IDE/Plugin/Completions/Logic.hs#L907-L908